monitor.js

```diff
@@ -0,0 +1,94 @@
+const pull = require('pull-stream')
+
+function isForeignLeafFeedBranch(branch) {
+  if (branch.length !== 4) return false
+  if (branch[0].keys) return false // not interested in my branches
+  if (branch[0].purpose !== 'root') return false
+  if (branch[1].purpose !== 'v1') return false
+  if (branch[2].purpose.length !== 1) return false
+  return true
+}
+
+/**
+ * There may have been messages I had in my database but was unable to decrypt
+ * because I didn't know my root metafeed keys. Now that I know them, I should
+ * reindex my database and decrypt all messages.
+ *
+ * Then, we should restart monitoring because we have the root and now we can
+ * finally create DM keys for foreign leafs.
+ */
+function reindexAndRestart(ssb, encryptionFormat, drain) {
+  if (drain) drain.abort()
+  ssb.db.reindexEncrypted((err) => {
+    if (err) console.warn(err)
+
+    monitorForest(ssb, encryptionFormat)
+  })
+}
+
+/**
+ * Listen to a stream of all branches in the metafeed "forest" (collection of
+ * trees) and send some information such as keys to the box2 encryptionFormat.
+ *
+ * Box2 encryptionFormat uses keys from the metafeed tree to encrypt messages,
+ * this is why we need this information
+ */
+function monitorForest(ssb, encryptionFormat) {
+  encryptionFormat.getRootSigningKey((err, rootKeys) => {
+    if (err) console.warn(err)
+
+    let drain = null
+    pull(
+      ssb.metafeeds.branchStream({ old: true, live: true }),
+
+      // For all feeds that I own, add their keys to the box2 encryptionFormat
+      pull.asyncMap((branch, cb) => {
+        let foundMyRoot = false
+        for (let i = 0; i < branch.length; i++) {
+          const { id, keys, purpose } = branch[i]
+          if (keys) {
+            encryptionFormat.addSigningKeys(id, keys)
+            if (i === 0 && purpose === 'root' && !rootKeys) {
+              foundMyRoot = encryptionFormat.addSigningKeys('root', keys)
```

Comment:
this addSigningKeys function doesn't return anything, or am I not reading this right?

Reply:
Oh yeah this is an effect of the ssb-keyring PR not being up yet. That function returns a boolean, where true means "it was not there before, this is new" and false means "I didn't add it because it was already there".

Reply:
Should be documented now

Reply:
Oh, my bad, this function is in this module and indeed doesn't return anything

Reply:
Fixed, now it should really return a boolean